torch/csrc/jit/codegen/cuda/python_frontend/README.md

```diff
@@ -0,0 +1,119 @@
+* `get()`: Provides a global instance.  You cannot directly construct an instance of the `FusionManager`.
+* `execute([list of inputs])`:  Allows you to execute the currently defined fusion with a list of given inputs.
+* `print_ir()`: Prints the low level IR for the currently defined fusion.
+
+### `FusionDefiniton` Context Manager - Interface for Defining Fusions
+
+#### Defining Input Tensors
+_All intermediate tensors are created by operations.  Constant tensors do not exist._
+
+There are 3 ways to define tensors that will be enumerated below.
+
+##### 1.) Defining tensors by the number of input dimensions only
+This interface tells nvFuser that the tensor has a given number of symbolic dimensions that are not necessarily contiguous in memory.  The user also has the ability to specify a data type.  The default type is `Float`.
+```
+t0 = fd.define_tensor(3)
+t1 = fd.define_tensor(3, DataType.Half)
+```
+
+##### 2.) Defining tensors by a list of concrete sizes and a list of strides
+The `sizes` parameter defines the number of dimensions and the size of each dimension.  The `strides` parameter has to have the same number of dimensions as the `sizes` parameter.
+nvFuser translates the concrete sizes and strides into symbolic sizes and contiguity information that can be directly defined via the next way to define tensors.  This allows the user to directly take a Pytorch defined tensor and query its sizes and strides in order to apply them in the definition.
+```
+t0 = fd.define_tensor(sizes=[2, 4, 6], strides=[24, 6, 1], dtype=DataType.Half)
+```
+
+##### 3.) Defining tensors by a list of symbolic sizes and a list of contiguity information
+The list of symbolic sizes defines the number of dimensions and `-1` is given for each dimension unless it is a broadcast dimension that is defined with a `1`.  The contiguity information is viewed from right to left.  A `True` definition indicates the current dimension is contiguous with the dimension to its right.
+
+```
+t0 = fd.define_tensor(symbolic_sizes=[-1, 1, -1], contiguous=[True, True, True], dtype=DataType.Float)
+```
+
+#### Defining Input Scalars
+_All intermediate scalars, except for constants, are created by operations._
+
+The only thing the user has to define for a scalar is its type.
+
+```
+s0 = fd.define_scalar(dtype=DataType.Half)
+```
+
+#### Defining Constant Scalars
+
+Constants can be of types: `Bool`, `ComplexDouble`, `Double`, or `Int`.  The definition only takes a constant and the type is inferred by the constant.
+
+```
+c0 = fd.define_constant(3.0)
+```
+
+#### Defining Operations
+
+Operators are added with the following notation:
+```
+output = fd.ops.foo(arg1, ... )
+```
+You can see a supported list of operations with the following query:
+```
+python -c "from torch._C._nvfuser import FusionDefinition; help(FusionDefinition.Operators)"
+```
+#### Notating Outputs
+
+`fd.add_output(tensor/scalar)` is used to indicate an intermediate is an output to the fusion.
```

Comment:
novice question: does `tensor/scalar` mean a division? or does this mean that `add_output` method can take one tensor or scalar?

Reply:
I meant one tensor or one scalar.  Would `fd.add_output(tensor|scalar)` be clearer?

Reply:
Python type annotation would look clearer to me.
e.g. `fd.add_output(output: typing.Union[torch.Tensor, numbers.Number])`

Reply:
Since it isn't a union, I just enumerated the options 

```python
add_output(output: Tensor)
# or
add_output(output: Scalar)
```